state-chain/chains/src/sol/instruction_builder.rs

```diff
@@ -340,13 +343,38 @@ impl SolanaInstructionBuilder {
 			sys_var_instructions(),
 		).with_remaining_accounts(ccm_accounts.remaining_account_metas())];
 
-		// TODO: Complete in PRO-1479
 		Self::finalize(
 			instructions,
 			nonce_account.into(),
 			agg_key.into(),
 			compute_price,
-			COMPUTE_LIMIT,
+			Self::calculate_gas_limit(gas_budget, compute_price, SolAsset::SolUsdc),
+		)
+	}
+
+	fn calculate_gas_limit(
+		gas_budget: SolAmount,
+		compute_price: SolAmount,
+		asset: SolAsset,
+	) -> SolComputeLimit {
+		let budget_after_signature = gas_budget.saturating_sub(LAMPORTS_PER_SIGNATURE);
+		if compute_price == 0 {
+			return DEFAULT_COMPUTE_UNITS_PER_CCM_TRANSFER;
+		}
+		let compute_budget =
+			// Budget is in lamports, compute price is in microlamports/CU
+			sp_std::cmp::min(
+				MAX_COMPUTE_UNITS_PER_TRANSACTION as u128,
+				(budget_after_signature as u128 * MICROLAMPORTS_PER_LAMPORT as u128)
+					/ (compute_price as u128),
+			) as SolComputeLimit;
```

Comment:
what is the difference between the gas_budget provided while constructing the tx and the hardcoded limit of ```MAX_COMPUTE_UNITS_PER_TRANSACTION```? arent both limitation set by us? If yes, why do we need two different limtis?

Reply:
`gas_budget` is in a CCM the amount of native token on the egress chain (SOL in this case) that the user has paid for. In a CCM we should never spend more than that amount in fees.
`MAX_COMPUTE_UNITS_PER_TRANSACTION` is the maximum amount of CUs that a transaction can consume, which is a limit set by the SVM (Solana), not by us.

Reply:
ok, so even if the user pays for a very a high budget, we cant have the budget implied compute units exceed the hardcoded limit? So in case that turns out to be the case, I guess we just swallow the extra budget provided by the user? is that correct?

Reply:
yes exactly that :+1: